lib/srv/transport/transportv1/transport.go

```diff
@@ -346,6 +353,42 @@ func getDestinationAddress(clientSrc, listenerAddr net.Addr) (net.Addr, error) {
 	}, nil
 }
 
+// checkedPrefixConn checks that read data has the specified prefix.
+type checkedPrefixConn struct {
```

Comment:
I didn't want to export some general checker, to keep it private, focused on one specific usage we control here. They are slightly different, this one wraps `net.Conn` and another one wraps `io.ReadWriteCloser`.

Reply:
I don't think this generics approach works, we get "Cannot use interfaces with methods in union" .

And generally I would agree that duplicated code has a danger of missing changes for one of the implementations, and it would be not suitable for general code in master, but since this version is only for this backport and it basically already has expiration date set (plus as mentioned earlier they are different a bit so there is even a complexity with implementing single one, if there's a way it's probably more complicated) I think it's ok for this case.

I've added comment as you suggested, mentioning that all changes should be mirrored.